americangut/diversity_analysis.py

```diff
@@ -0,0 +1,1729 @@
+    """Extracts the distance information for all samples in a group
+
+    Parameters
+    ----------
+    dm : skbio DistanceMatrix
+        A distance matrix object with the samples corresponding to those in
+        the data frame with the metadata.
+    df : pandas DataFrame
+        A dataframe containing the metadata associated withed with the object
+    group : str
+        the metadata category being interograted, used to group the data
+    order : None, list
+        Default is None. The order of groups in group.
+
+    Returns
+    -------
+    within : list
+        A list of the group names for the within group distance
+    w_dist : list
+        A list of arrays with the within group distances
+    between : list
+        A list of the group names for between group distance
+    b_dist : list
+        A list of arrays for between group distance
+
+    """
+    # Checks the column is supported
+    if group not in df:
+        raise ValueError('%s is not a metadata category.' % group)
+
+    # Gets the sample ids associated with the group
+    if order is None:
+        order = list(df.groupby(group).groups)
+    ordered_ids = {o: df.groupby(group).groups[o] for o in order}
+
+    # Gets the data
+    dist_ids = np.array(list(dm.ids))
+    dist_data = dm.data
+
+    # Prealocates objects for return
+    within = []
+    w_dist = []
+    between = []
+    b_dist = []
+
+    # Loops through the groups
+    for id1, o1 in enumerate(order):
+        # Gets the intragroup distance
+        within.append(o1)
+        w_dist.append(dm.filter(ordered_ids[o1]).condensed_form())
+        for o2 in order[(id1+1):]:
+            loc1 = np.array([dist_ids == i for i in ordered_ids[o1]]).any(0)
+            loc2 = np.array([dist_ids == i for i in ordered_ids[o2]]).any(0)
+            data1 = dist_data[loc1, :]
+            # Adds the intragroup distance
+            between.append({o1, o2})
+            b_dist.append((data1[:, loc2]).flatten())
+
+    return within, w_dist, between, b_dist
+
+
+def beta_diversity_bars(dm, meta, group, order=None, ref_groups=None,
+    num_iter=999, p_crit=0.01, p_table=None,
+    p_tab_col='Parametric p-value (Bonferroni-corrected)',
+    ref_less=True, ax=None, interval=0.1, width=0.1,
+    show_seperation=True, colormap=None, match_colors=True,
+    elinewidth=2, ecapwidth=2, show_p=False, lowest=None,
+    sep_size=0.035, label_size=12, show_p_value=False, **kwargs):
+    """Creates a barchart of the beta diversity distances
+
+    Parameters
+    ----------
+    dm : skbio DistanceMatrix
+        A distance matrix object with the samples corresponding to those in
+        the data frame with the metadata.
+    meta : pandas dataframe
+        The metadata associated with the samples in the DataFrame. The samples
+        in the DataFrame may be a superset of the samples in the distance
+        matrix, `dm`.
+    group : str
+        the metadata category being interograted, used to group the data
```

Comment:
interrogated

again, copy and paste is dangerous but for more than just spelling